base/indices.jl

```diff
@@ -394,9 +394,52 @@ last(S::IdentityUnitRange) = last(S.indices)
 size(S::IdentityUnitRange) = (length(S.indices),)
 length(S::IdentityUnitRange) = length(S.indices)
 unsafe_length(S::IdentityUnitRange) = unsafe_length(S.indices)
-getindex(S::IdentityUnitRange, i::Int) = (@_inline_meta; @boundscheck checkbounds(S, i); i)
-getindex(S::IdentityUnitRange, i::AbstractUnitRange{<:Integer}) = (@_inline_meta; @boundscheck checkbounds(S, i); i)
-getindex(S::IdentityUnitRange, i::StepRange{<:Integer}) = (@_inline_meta; @boundscheck checkbounds(S, i); i)
+function getindex(S::IdentityUnitRange, i::Integer)
+    @_inline_meta
+    i isa Bool && throw(ArgumentError("invalid index: $i of type Bool"))
+    @boundscheck checkbounds(S, i)
+    i
+end
+function getindex(S::IdentityUnitRange, i::AbstractUnitRange{<:Integer})
+    @_inline_meta
+    @boundscheck checkbounds(S, i)
+    if eltype(i) === Bool
+        # logical indexing
+        if length(i) == 0
+            return range(first(S), length = 0)
+        elseif length(i) == 1
+            if first(i)
+                return range(first(S), length = length(S))
+            else
+                return range(first(S), length=0)
+            end
+        else # length(i) == 2
+            return range(first(S) + one(first(S)), length=1)
+        end
+    else
+        return i
```

Comment:
Not sure about this one — but perhaps it should be:

```suggestion
        return Int.(i)
```

Reply:
Thanks for the suggestions! Perhaps the correct choice here is `map(Int, i)` as it has special methods defined to return ranges, whereas broadcasting currently converts a range to a `Vector`.

```julia
julia> Int.(big(2):big(3))
2-element Vector{Int64}:
 2
 3

julia> map(Int, big(2):big(3))
2:3
```

I'll create a new commit with all the changes incorporated.